### PR TITLE
[ML] Adjust finalize job action to work with documents

### PR DESCRIPTION
**Pull request by davidkyle**
For the JIndex feature branch.

The Finalize job action sets the job finished time and should be called only when the job is closed. The change is fairly simple: add an internal finished time field to JobUpdate and update the job in the action. 

I take it as a sign of progress that this change can cause Version conflict exceptions between `AutoDetectResultProcessor` and the finalize action as both attempt to update the job. I marked this WIP as I attempt to resolve the problem some options are:
1. Reduce contention. AutoDetectResultProcessor is trying to set established model memory could this be stored somewhere else?
2. Retry on failure.
3. 1 and 2

### CODE REVIEW

**Comment by droberts195** [comment]
++ to

> 3. 1 and 2

Due to #34084 `established_model_memory` can be removed from the `Job` object.  (Parsing must tolerate it for BWC, but it doesn't need to be stored or written.)

**Comment by droberts195** [comment]
> After this change there was no need for the `FinalizeJobAction` so I removed it but this may be a bad idea in a mixed cluster as a closing job on an old node will try to call that action and it no longer exists. I'm not sure what should be done here, maybe retain `FinalizeJobAction` and make it a no-op.

I think we need to know what the effect of this is in a mixed node cluster.  If the older version node still closes the job it may be OK as it is.  If the older node suffers an exception that causes the job to go into the `failed` state or returns a scary exception to the end user then we'll need to keep the `FinalizeJobAction` around as a no-op until 7.last.

**Review by droberts195 — COMMENTED** [review]
I left a couple of questions about BWC wire formats and inconsistencies in where `isInternal` has been removed.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/UpdateJobAction.java [hunk 1]
```diff
@@ -111,11 +102,6 @@ public void readFrom(StreamInput in) throws IOException {
             } else {
                 isInternal = false;
             }
-            if (in.getVersion().onOrAfter(Version.V_6_3_0)) {
-                waitForAck = in.readBoolean();
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/UpdateJobAction.java]: I don't know what I was thinking I should have reviewed this before asking for a review sorry. I pushed 2 more commits one addresses this and the other adds the finalize job action back into the security tests.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**